src/lib.rs

```diff
@@ -1,8 +1,84 @@
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        let result = 2 + 2;
-        assert_eq!(result, 4);
+//! # Dotium
+//!
+//! The uniform object provider for Minecraft Mod hosting platforms.
+//! Implements traits for api wrappers Ferinth, Furse, and Octocrab to then be used platform independedly in GDLauncher
+
+#[cfg(feature = "curseforge")]
+mod curseforge;
+#[cfg(feature = "github")]
+mod github;
+#[cfg(feature = "modrinth")]
+mod modrinth;
+pub mod project;
+pub mod request;
+pub mod version;
+
+use async_trait::async_trait;
+use project::{Author, Project};
+use serde::{Deserialize, Serialize};
+use version::Version;
+
+#[derive(Debug, Clone)]
+pub struct Platform {
+    /// The name of the platform
+    pub name: String,
+    /// The two-letter nick name of the Platform
+    pub short_name: String,
+    /// Logo of the platform
+    pub logo: request::Asset,
+}
+
+#[derive(thiserror::Error, Debug)]
+#[error("The trait is not implemented")]
+pub struct UnimplementedError;
+
+pub(crate) type Datetime = chrono::DateTime<chrono::Utc>;
+
+#[async_trait]
+pub trait Container<E: From<UnimplementedError>, ID: Sync + Send> {
+    fn new() -> Self;
+    fn get_platform() -> Platform;
+
+    // GATs unstable
+    // type Result<T> = std::result::Result<T, E>;
+
+    async fn get_project(&self, id: &ID) -> Result<Project<ID>, E>;
+
+    async fn get_projects(&self, ids: Vec<&ID>) -> Result<Vec<Project<ID>>, E> {
+        let mut projects = Vec::new();
+        for id in ids {
```

Comment:
Not totally sure about that, but wouldn't this mean that we are working a project at a time?
Could something like https://docs.rs/futures/latest/futures/future/fn.join_all.html help?

Reply:
Sometimes too many concurrent requests completely fails, I had to deal with this problem when implementing concurrent requests, and it was a rather huge pain. I ended up using `Semaphore` which I don't think you can use with `join_all()`, and semaphore would need to import tokio which we certainly don't want.

Reply:
Seems like we have the first design problem lol
But well, it's not like it is a huge problem, and the solution (Having pending requests we can start at any time) is even worse

Reply:
Well the only platform that doesn't implement multiple requests is github (the others overrides this default implementation), and we won't be using the multiple requests on github afaik.